In [17]:
import main.dsl.mathnum.*

In [18]:
run {
    val m = Variable("m", "slope").setTo(1)
    val x = Variable("x").setTo(3)
    val b = Variable("b", "y intercept").setTo(0)
    
    // usage as a math function 
    val y = MathFunction(m * x + b, "y")

    y.display()
}

[Variable(glyph=m, value=1, name=slope, instance=null, isBound=false), Variable(glyph=x, value=3, name=null, instance=null, isBound=false), Variable(glyph=b, value=0, name=y intercept, instance=null, isBound=false)]


y = (b,m,x) \mapsto m \cdot x + b = 3.0

In [19]:
run {
    val m = Variable("m", "slope").setTo(1)
    val x = Variable("x").setTo(3)
    val b = Variable("b", "y intercept").setTo(0)

    // usage as an equation
    val y = Variable("y").standsFor(m * x + b)

    y.explain()
}


Given
b = 0 (y intercept)
m = 1 (slope)
x = 3
y = 3.0

y = m \cdot x + b

yields (3.0, 3.0)

$$
y = m \cdot x + b
$$

In [21]:
run {
    val E = Variable("E", "Young's Modulus")
    val A = Variable("A", "Cross sectional area")
    val L = Variable("L")

    val springLinear = MathFunction((E * A) / L)

    E.setTo(23)
    A.setTo(35)
    L.setTo(22)
    springLinear.evaluate()
}

36.59090909090909

In [27]:
val v = Variable("v").setTo(17.0)
val e = 7.num() + v + 13.0 / 32.0 + 17.num().squared()

(0..5).forEach {
    v.setTo(it)
    e.evaluate().also { println(it) }
}

println("==============")
e.toLatex().also { println(it) }
println("==============")


296.40625
297.40625
298.40625
299.40625
300.40625
301.40625
7 + v + 0.40625 + 17^{2}


In [28]:
    e.toGraphviz().also{println(it)}

            digraph {
            
            rankdir = "RL"
           
            "Add\n1713561422" -> "Add\n159186657" [ label="1" ]
"Add\n1713561422" -> "Exponent\n1073742351" [ label="2" ]
"Add\n159186657" -> "Add\n109125887" [ label="1" ]
"RealNum\n(0.40625)\n74784828" [style=diagonals, color="#0b008c"]
"Add\n159186657" -> "RealNum\n(0.40625)\n74784828" [ label="2" ]
"IntegerNum\n(7.0)\n-1256214987" [style=diagonals, color="#0b008c"]
"Add\n109125887" -> "IntegerNum\n(7.0)\n-1256214987" [ label="1" ]
"Variable\n'v' (5.0)\n-2005624170" [style=diagonals, color="#ff0000"]
"Add\n109125887" -> "Variable\n'v' (5.0)\n-2005624170" [ label="2" ]
"IntegerNum\n(17.0)\n2133108511" [style=diagonals, color="#0b008c"]
"Exponent\n1073742351" -> "IntegerNum\n(17.0)\n2133108511" [ label="1" ]
"RealNum\n(2.0)\n2024750116" [style=diagonals, color="#0b008c"]
"Exponent\n1073742351" -> "RealNum\n(2.0)\n2024750116" [ label="2" ]


            }            


            digraph {
            
            rankdir = "RL"
           
            "Add\n1713561422" -> "Add\n159186657" [ label="1" ]
"Add\n1713561422" -> "Exponent\n1073742351" [ label="2" ]
"Add\n159186657" -> "Add\n109125887" [ label="1" ]
"RealNum\n(0.40625)\n74784828" [style=diagonals, color="#0b008c"]
"Add\n159186657" -> "RealNum\n(0.40625)\n74784828" [ label="2" ]
"IntegerNum\n(7.0)\n-1256214987" [style=diagonals, color="#0b008c"]
"Add\n109125887" -> "IntegerNum\n(7.0)\n-1256214987" [ label="1" ]
"Variable\n'v' (5.0)\n-2005624170" [style=diagonals, color="#ff0000"]
"Add\n109125887" -> "Variable\n'v' (5.0)\n-2005624170" [ label="2" ]
"IntegerNum\n(17.0)\n2133108511" [style=diagonals, color="#0b008c"]
"Exponent\n1073742351" -> "IntegerNum\n(17.0)\n2133108511" [ label="1" ]
"RealNum\n(2.0)\n2024750116" [style=diagonals, color="#0b008c"]
"Exponent\n1073742351" -> "RealNum\n(2.0)\n2024750116" [ label="2" ]


            }            


$$7 + 0.40625 + 17^{2}$$

## Using a MathFunction to verify that our algorithm is set up correctly

Trying to implement the formula which is defined [here](https://math.stackexchange.com/questions/154941/how-to-calculate-the-number-of-possible-connected-simple-graphs-with-n-labelle)

 

In [24]:
import main.dsl.expressions.ScalarAlgebra

val n = Variable("n").setTo(0)
val x = Variable("x").setTo(5)
val connectedGraphsWithXVertices = MathFunction(
    ScalarAlgebra.Sum(
        n,
        mathNum(0),
        mathNum(10000),
        {
            val numerator = 2.0.num() pow (n * (n - 1) / 2)
            (numerator / ScalarAlgebra.Factorial(n)) * (x pow n)
        }),
    "D(x)"
)

connectedGraphsWithXVertices.evaluate().also { println(it) }
connectedGraphsWithXVertices.display()


10001.0
[Variable(glyph=n, value=0, name=null, instance=null, isBound=false)]


D(x) = (n) \mapsto \sum_{n=0}^{10000}{\frac{2^{\frac{n \cdot (n - 1)}{2}}}{n!} \cdot x^{n}} = 10001.0

$$
D(x) = (n) \mapsto \sum_{n=0}^{10000}{\frac{2^{\frac{n \cdot (n - 1)}{2}}}{n!} \cdot x^{n}} = NaN
$$